### DISCUSSION TITLE
Best way to refresh data after useContractWrite

### DISCUSSION THREAD

**Discussion by Fabien-Morrow** [wevm/wagmi#1238] [General]
Hi,
I'm showing some proposals on a dashboard stored on a smartcontract using useContractRead, or useContractInfiniteRead.
I've also an option to create a new proposal, using usePrepareContractWrite and useContractWrite under the hoods.

What is the best way to refresh my proposals after creating a new one ? I've tried severals options and I'm not sure what's the best solution :

1) Emit NewProposal in the smartcontract when a new proposal is created, then catching it and refresh my data with
```js
  const {
    data: proposals,
    refetch,
  } = useContractInfiniteReads({...mysetup});
// (...)
  useContractEvent({
    ...contractConfig,
    eventName: "NewProposal",
    listener: (event) => refetch(),
  });
```
Pros : On the paper it's perfect, it's elegant.
Cons : It simply doesn't work, because generaly the provider haven't got the last block yet, and I basically refetch stale data :/

2) Same with above but with
```js
  useContractEvent({
    ...contractConfig,
    eventName: "NewProposal",
    listener: (event) => setTimeout(refetch, 5000),
  });
```
Cons : (if provider is slow) Sometimes it doesn't work ! Also I don't feel good when I write this.
Pros : (if provider is fast) Sometimes it works !

3) I use useWaitForTransaction after useContractWrite with a confirmation of 2 or 3, launching refetch

```js
  const { refetch } = useMyCustomHookContainingUseContractInfiniteReads();
  const { data, write } = useContractWrite(config);
  const { isLoading, status } = useWaitForTransaction({
    confirmations: 2,
    hash: data?.hash,
    onSuccess(data) {
      refetch();
    },
```
Cons : (if provider is slow) Sometimes it doesn't work ! Even with 3 confirmations :/
Pros : (if provider is fast) Sometimes it works !

Do you have others approaches ? What's your advises on this ?

**Comment by Fabien-Morrow** [comment]
I'm quite tired, I just realise, in 1) : If I catch the event, it's the provider who provided me this event I'm right ? So he reads it in the log of the block I'm right ? So the provider is in possession of the block containing the updated state of my contract I'm right ?

So why the date returned is stale ?

Is my code somewhere wrong ?

**Comment by fachrudahri** [comment]
in useContractRead you need added `watch: true`, and place data from useContractRead in dependencies array useEffect and wrap the refetch with the useEffect
